src/FrequencyBalancerAgent.cpp

```diff
@@ -0,0 +1,644 @@
+/*
+ * Copyright (c) 2015 - 2023, Intel Corporation
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+#include "config.h"
+
+#include "FrequencyBalancerAgent.hpp"
+
+#include <unistd.h>
+
+#include "Environment.hpp"
+#include "FrequencyGovernor.hpp"
+#include "FrequencyLimitDetector.hpp"
+#include "FrequencyTimeBalancer.hpp"
+#include "PowerGovernor.hpp"
+#include "SSTClosGovernor.hpp"
+#include "Waiter.hpp"
+#include "geopm/Exception.hpp"
+#include "geopm/Helper.hpp"
+#include "geopm/PlatformIO.hpp"
+#include "geopm/PlatformTopo.hpp"
+#include "geopm_debug.hpp"
+#include "geopm_hash.h"
+#include "geopm_hint.h"
+#include "geopm_time.h"
+#include "geopm_topo.h"
+
+#include <algorithm>
+#include <chrono>
+#include <cmath>
+#include <exception>
+#include <iterator>
+#include <numeric>
+#include <thread>
+#include <utility>
+
+// Minimum number of sampling wait periods before applying new epoch controls.
+#define MINIMUM_WAIT_PERIODS_FOR_NEW_EPOCH_CONTROL 5
+
+// Minimum number of epochs to wait before applying new epoch controls.
+#define MINIMUM_EPOCHS_FOR_NEW_EPOCH_CONTROL 3
+
+// Number of back-to-back network hints to treat as "in a network region."
+// Lower numbers respond more quickly, but risk throttling regions that
+// happen to land next to a short-running network region.
+// -- Arbitrarily set to 3 because that produces acceptable behavior so far.
+#define NETWORK_HINT_MINIMUM_SAMPLE_LENGTH 3
+
+// Number of back-to-back non-network hints to treat as not "in a network region."
+#define NON_NETWORK_HINT_MINIMUM_SAMPLE_LENGTH 1
+
+namespace geopm
+{
+    FrequencyBalancerAgent::FrequencyBalancerAgent()
+        : FrequencyBalancerAgent(platform_io(), platform_topo(),
+                                 Waiter::make_unique(environment().period(M_WAIT_SEC)),
+                                 PowerGovernor::make_shared(),
+                                 FrequencyGovernor::make_shared(),
+                                 (SSTClosGovernor::is_supported(platform_io())
+                                      ? SSTClosGovernor::make_shared()
+                                      : nullptr),
+                                 {}, {})
+    {
+    }
+
+    static bool is_all_nan(const std::vector<double> &vec)
+    {
+        return std::all_of(vec.begin(), vec.end(),
+                           [](double x) -> bool { return std::isnan(x); });
+    }
```

Comment:
I see that this is replicated here, the FrequencyMapAgent, and the FFNet agent.  IMO it should really be in service/src/Helper.cpp instead of being replicated again.